File: solutions/clojure/cars-assemble/1/src/cars_assemble.clj
(ns cars-assemble)

(def cars-per-hour 221)

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
    (cond (= speed 0) 0
      (<= speed 4) (* speed cars-per-hour)
      (<= speed 8) (* speed cars-per-hour 0.90)
      (= speed 9) (* speed cars-per-hour 0.80)
      (= speed 10) (* speed cars-per-hour 0.77)
    )
  )

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (quot (production-rate speed) 60))
  )

File: solutions/go/cars-assemble/2/cars_assemble.go
package cars

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * (successRate / 100.0)
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	carsPerHour := CalculateWorkingCarsPerHour(productionRate, successRate)
	const minutesPerHour = 60.0
	return int(carsPerHour / minutesPerHour)
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	const (
		carsPerGroup = 10
		costPerCar   = 10_000
		costPerGroup = 95_000
	)

	carGroups := carsCount / carsPerGroup
	carsRemaining := carsCount % carsPerGroup

	return uint(carGroups*costPerGroup + carsRemaining*costPerCar)
}
